### ISSUE TITLE
Potential bug on Ubuntu 14

### ISSUE THREAD

**Issue by dwmkerr** [dwmkerr/node-docker-microservice#4]
It seems there is a bug affecting some people, see:

http://www.dwmkerr.com/learn-docker-by-building-a-microservice/#comment-2716205460

For details.

Possible things to check:
1. Docker version
2. OS version
3. MySQL version

**Comment by EliasJorgensen** [comment]
When i run start.sh i get the following message:

> Starting DB...
> 2a04c50aae663c52bb6cdf5f8ca3516a9e32b016c018854c8b01b26a03b73812
> Waiting for DB to start up...
> mysqladmin: [Warning] Using a password on the command line interface can be insecure.
> Got error: Access denied for user 'users_service'@'localhost' (using password: YES)
> Setting up initial data...
> mysql: [Warning] Using a password on the command line interface can be insecure.
> ERROR 1045 (28000): Access denied for user 'users_service'@'localhost' (using password: YES)

I'm on Ubuntu 16.04 LTS.
Docker version 1.11.2, build b9f10c9.
Latest version of the MySQL image.

**Comment by rohanliston** [comment]
Same thing here.

Ubuntu 16.04 LTS
Docker version 1.11.1, build 5604cbe
mysql:latest

**Comment by ElijahLynn** [comment]
Same here, when I checkout step1 or step2. 

If I run the start command manually on the CLI it does work. It seems related to the environment variables not being ready in time. 

Docker version 1.7.1, build 786b29d 

![image](https://cloud.githubusercontent.com/assets/1504756/16397056/22128348-3c90-11e6-996a-3faada4cb79c.png)

**Comment by dwmkerr** [comment]
Hi @ElijahLynn interesting that you mention the environment variables - what makes you think they might not be ready in time? A timing issue could indeed be the root cause as it seems inconsistent and doesn't affect many users

**Comment by ElijahLynn** [comment]
Hi @dwmkerr, I say that because after the container is started, the users_service user is present and can login with the password specified. At least that is what I remember a week or so ago. I will double check.

**Comment by ElijahLynn** [comment]
Yes, immediately after I get the error messages in https://github.com/dwmkerr/node-docker-microservice/issues/4#issuecomment-228887482 I run `docker exec -it db bash` && `mysql -p123` && `select user from mysql.users;` and I get two users, root and users_service.

I also just upgraded to Docker version 1.11.2, build b9f10c9.

![image](https://cloud.githubusercontent.com/assets/1504756/16414063/7e888546-3d03-11e6-943a-fd28f8eac22b.png)

**Comment by ElijahLynn** [comment]
I can also run the command in the script `docker exec db mysqladmin --wait=30 -uusers_service -p123 ping` immediately after and I get 'mysqld is alive'.

**Comment by ElijahLynn** [comment]
Okay, so it appears that if we add a `sleep 10` right after the run command then the script works. Worth noting that <= 9 seconds it fails.

**Comment by ElijahLynn** [comment]
Maybe some systems are just faster at initializing the container and MySQL and that is why some have the issue and some don't. Odd though, I am on a pretty fast system, 8 core, 32GB ram, Intel SSD etc. Maybe my config is not optimal if this is the culprit.

**Comment by dwmkerr** [comment]
Well done for finding a solution @ElijahLynn - I'll update the write-up and code

**Comment by ElijahLynn** [comment]
Cool, it may not work on some systems that are slower, maybe put a comment in there saying some may need to increase the sleep if they get access denied errors.

**Comment by oschrenk** [comment]
Not specific to Ubuntu. This also happens on Docker for Mac 1.12. Adding the sleep command helps.

**Comment by mdender** [comment]
@ElijahLynn thanks, I'm on OSX and adding that sleep solved the issue for me.

**Comment by That-David-Guy** [comment]
Bugs out for me on step 2. Sleep doesn't do anything. This is what I've added to my start.sh. (I'm on a macbook air)

**Comment by That-David-Guy** [comment]
```
...
echo "Sleep to fix race condition"
docker exec db sleep 10
...
```

**Comment by That-David-Guy** [comment]
Correction, `docker exec db sleep 30` worked. 10 seconds mustn't of been enough for me :(
